merkle.go

```diff
@@ -124,25 +124,31 @@ func (t *Tree) AddLeaf(value []byte) error {
 		// If no node is pending, then this node is a left sibling,
 		// pending for its right sibling before its parent can be calculated.
 		if l.parking.IsEmpty() {
-			l.parking = n
+			// Copy the byte slice as we will keep it for a while.
+			l.parking.value = append(l.parking.value[:0], n.value...)
+			l.parking.OnProvenPath = n.OnProvenPath
 			break
 		} else {
 			// This node is a right sibling.
-			lChild, rChild = l.parking, n
-			parent = t.calcParent(lChild, rChild)
+			lChild, rChild := l.parking, n
 
 			// A given node is required in the proof if and only if its parent is an ancestor
 			// of a leaf whose membership in the tree is being proven, but the given node isn't.
-			if parent.OnProvenPath {
+			if lChild.OnProvenPath || rChild.OnProvenPath {
 				if !lChild.OnProvenPath {
-					t.proof = append(t.proof, lChild.value)
+					copy := append([]byte(nil), lChild.value...)
+					t.proof = append(t.proof, copy)
 				}
 				if !rChild.OnProvenPath {
-					t.proof = append(t.proof, rChild.value)
+					copy := append([]byte(nil), rChild.value...)
+					t.proof = append(t.proof, copy)
 				}
 			}
 
-			l.parking.value = nil
+			parent := t.calcParent(t.parentBuf[:0], lChild, rChild)
+			t.parentBuf = parent.value
```

Comment:
I didn't use `t.parentBuf` in `RootAndProof` as it complicated it without any perf improvement. The calculated parent needed to be copied anyway in every code path.

Reply:
1. What I meant is that inside the implementation of `t.calcParent` you can pass `t.parentBuf` into `t.hash` directly, instead of accepting it as an argument.
2. Why do this `t.parentBuf = parent.value`? Why not just delete this line? (it's now `t.parentBuf = n.value`)

Reply:
Ad 1.
Because I don't want to reuse `t.parentBuf` in `RootAndProof` method (hence it passes `nil`). It only complicated the code there.

Ad 2. 
It needs to reassign because the buffer passed into `t.hash()` will be reallocated the first time it is called (we don't know the size of the required buffer upfront to preallocate it or even use an array). It's the same story as with `buf = append(buf, ...)`.